pkg/service_sync/service_sync.go

```diff
@@ -0,0 +1,266 @@
+	localServices     map[string]types.ServiceInterface
+	byName            map[string]types.ServiceInterface
+	heardFrom         map[string]time.Time
+}
+
+type ServiceUpdate struct {
+	origin      string
+	version     string
+	definitions map[string]types.ServiceInterface
+}
+
+func NewServiceSync(origin string, version string, connectionFactory messaging.ConnectionFactory, handler UpdateHandler) *ServiceSync {
+	s := &ServiceSync{
+		origin:            origin,
+		version:           version,
+		handler:           handler,
+		connectionFactory: connectionFactory,
+		outgoing:          make(chan ServiceUpdate, 10),
+		incoming:          make(chan ServiceUpdate, 10),
+		updates:           make(chan map[string]types.ServiceInterface, 10),
+		byOrigin:          map[string]map[string]types.ServiceInterface{},
+		localServices:     map[string]types.ServiceInterface{},
+		byName:            map[string]types.ServiceInterface{},
+		heardFrom:         map[string]time.Time{},
+	}
+	return s
+}
+
+func (c *ServiceSync) LocalDefinitionsUpdated(definitions map[string]types.ServiceInterface) {
+	c.updates <- definitions
+}
+
+func (c *ServiceSync) pareByOrigin(service string) {
+	for _, origin := range c.byOrigin {
+		if _, ok := origin[service]; ok {
+			delete(origin, service)
+			return
+		}
+	}
+}
+
+func (c *ServiceSync) recordChanges(definitions map[string]types.ServiceInterface) {
+	var added []types.ServiceInterface
+	var modified []types.ServiceInterface
+	var removed []types.ServiceInterface
+
+	for _, def := range c.byName {
+		latest, ok := definitions[def.Address]
+		if !ok {
+			removed = append(removed, def)
+		} else if !equivalentServiceDefinition(&def, &latest) {
+			modified = append(modified, def)
+		}
+	}
+	for _, def := range definitions {
+		if _, ok := c.byName[def.Address]; !ok {
+			added = append(added, def)
+		}
+	}
+	if len(added) > 0 {
+		event.Recordf(ServiceSyncEvent, "Service interface(s) added %s", strings.Join(getAddresses(added), ","))
+	}
+	if len(removed) > 0 {
+		event.Recordf(ServiceSyncEvent, "Service interface(s) removed %s", strings.Join(getAddresses(removed), ","))
+	}
+	if len(modified) > 0 {
+		event.Recordf(ServiceSyncEvent, "Service interface(s) modified %s", strings.Join(getAddresses(modified), ","))
+	}
+}
+
+func (c *ServiceSync) localDefinitionsUpdated(definitions map[string]types.ServiceInterface) ServiceUpdate {
+	latest := make(map[string]types.ServiceInterface) // becomes c.localServices
+	byName := make(map[string]types.ServiceInterface)
+
+	for name, original := range definitions {
+		service := types.ServiceInterface{
+			Address:      original.Address,
+			Protocol:     original.Protocol,
+			Ports:        original.Ports,
+			Origin:       original.Origin,
+			Headless:     original.Headless,
+			Labels:       original.Labels,
+			Aggregate:    original.Aggregate,
+			EventChannel: original.EventChannel,
+			Targets:      []types.ServiceInterfaceTarget{},
```

Comment:
Missing `EnableTls` and `TlsCredentials`.

Reply:
I missed that, thank you